lib/backend/memory/memory.go

```diff
@@ -374,7 +378,69 @@ func (m *Memory) CompareAndSwap(ctx context.Context, expected backend.Item, repl
 	if !m.EventsOff {
 		m.buf.Emit(event)
 	}
-	return m.newLease(replaceWith), nil
+	return backend.NewLease(replaceWith), nil
+}
+
+func (m *Memory) ConditionalDelete(ctx context.Context, key []byte, rev string) error {
+	if len(key) == 0 {
+		return trace.BadParameter("missing parameter key")
+	}
+	m.Lock()
+	defer m.Unlock()
+	m.removeExpired()
+
+	item, found := m.tree.Get(&btreeItem{Item: backend.Item{Key: key}})
+	if !found {
+		return trace.NotFound("key %q is not found", string(key))
```

Comment:
CompareAndSwap always returns a CompareFailed even if the item doesn't exist; do we want ConditionalDelete and ConditionalUpdate to split the two cases?